### PR TITLE
[SPARK-47352][SQL] Fix Upper, Lower, InitCap collation awareness

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Add support for Locale aware expressions.

### Why are the changes needed?
This is needed as some future collations might use different Locales then default.

### Does this PR introduce _any_ user-facing change?
Yes, we follow ICU implementations for collations that are non native.

### How was this patch tested?
Tests for Upper, Lower and InitCap already exist.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 1]
```diff
@@ -261,6 +261,156 @@ public void testEndsWith() throws SparkException {
...
+    assertInitCap("AbĆdE","UNICODE", "Abćde");
+    assertInitCap("AbĆdE","UNICODE_CI", "Abćde");
+    // Variable byte length characters
+    assertInitCap("aB世De", "UTF8_BINARY", "Ab世de");
+    assertInitCap("ÄBĆΔE", "UTF8_BINARY", "Äbćδe");
+    assertInitCap("aB世De", "UTF8_BINARY_LCASE", "Ab世de");
+    assertInitCap("ÄBĆΔE", "UTF8_BINARY_LCASE", "Äbćδe");
+    assertInitCap("aB世De", "UNICODE", "Ab世de");
+    assertInitCap("ÄBĆΔE", "UNICODE", "Äbćδe");
+    assertInitCap("aB世De", "UNICODE_CI", "Ab世De");
```
Comment (mihailom-db) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:405]: @cloud-fan Should we just proceed with this ICU result? I am would expect Ab世de, but ICU seems to do conversion this way when we use UNICODE_CI.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 2]
```diff
@@ -144,6 +147,111 @@ public static boolean execICU(final UTF8String l, final UTF8String r,
...
+      } else if (collation.supportsLowercaseEquality) {
+        return String.format(expr + "Lowercase(%s)", v);
+      } else {
+        return String.format(expr + "ICU(%s, %d)", v, collationId);
+      }
+    }
+    public static UTF8String execBinary(final UTF8String v) {
+      return v.toUpperCase();
+    }
+    public static UTF8String execLowercase(final UTF8String v) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:175]: This is the same as `execBinary`?

**Review by cloud-fan — APPROVED** [review]

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 2 above)
Comment (mihailom-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java]: Merged these two to `execUTF8`.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**